File: simplified-tests/src/main/java/org/nypl/simplified/tests/opds/OPDSJSONParserContract.java
package org.nypl.simplified.tests.opds;

import org.junit.Assert;
import org.junit.Test;
import org.nypl.simplified.opds.core.OPDSAcquisition;
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry;
import org.nypl.simplified.opds.core.OPDSJSONParser;
import org.nypl.simplified.opds.core.OPDSJSONParserType;

import java.io.FileNotFoundException;
import java.io.InputStream;
import java.net.URL;

public abstract class OPDSJSONParserContract {

  private static InputStream getResource(
    final String name)
    throws Exception {

    final String path = "/org/nypl/simplified/tests/opds/" + name;
    final URL url = OPDSFeedEntryParserContract.class.getResource(path);
    if (url == null) {
      throw new FileNotFoundException(path);
    }
    return url.openStream();
  }

  @Test
  public void testCompatibility20180921_1()
    throws Exception {

    final OPDSJSONParserType json_parser =
      OPDSJSONParser.newParser();

    final OPDSAcquisitionFeedEntry e0 =
      json_parser.parseAcquisitionFeedEntryFromStream(
        OPDSJSONParserContract.getResource("compatibility-20180921-test-old.json"));

    final OPDSAcquisitionFeedEntry e1 =
      json_parser.parseAcquisitionFeedEntryFromStream(
        OPDSJSONParserContract.getResource("compatibility-20180921-test-new-1.json"));

    {
      Assert.assertEquals(e0.getAcquisitions(), e1.getAcquisitions());
      Assert.assertEquals(e0.getAvailability(), e1.getAvailability());
      Assert.assertEquals(e0.getAuthors(), e1.getAuthors());
      Assert.assertEquals(e0.getCategories(), e1.getCategories());
      Assert.assertEquals(e0.getCover(), e1.getCover());
      Assert.assertEquals(e0.getGroups(), e1.getGroups());
      Assert.assertEquals(e0.getID(), e1.getID());
      Assert.assertEquals(e0.getPublished(), e1.getPublished());
      Assert.assertEquals(e0.getPublisher(), e1.getPublisher());
      Assert.assertEquals(e0.getSummary(), e1.getSummary());
      Assert.assertEquals(e0.getThumbnail(), e1.getThumbnail());
      Assert.assertEquals(e0.getTitle(), e1.getTitle());
    }
  }

  @Test
  public void testCompatibility20180921_2()
    throws Exception {

    final OPDSJSONParserType json_parser =
      OPDSJSONParser.newParser();

    final OPDSAcquisitionFeedEntry e0 =
      json_parser.parseAcquisitionFeedEntryFromStream(
        OPDSJSONParserContract.getResource("compatibility-20180921-test-old.json"));

    final OPDSAcquisitionFeedEntry e1 =
      json_parser.parseAcquisitionFeedEntryFromStream(
        OPDSJSONParserContract.getResource("compatibility-20180921-test-new-0.json"));

    {
      final OPDSAcquisition e0a = e0.getAcquisitions().get(0);
      final OPDSAcquisition e1a = e1.getAcquisitions().get(0);

      Assert.assertEquals(e0a.availableFinalContentTypes(), e1a.availableFinalContentTypes());
      Assert.assertEquals(e0.getAvailability(), e1.getAvailability());
      Assert.assertEquals(e0.getAuthors(), e1.getAuthors());
      Assert.assertEquals(e0.getCategories(), e1.getCategories());
      Assert.assertEquals(e0.getCover(), e1.getCover());
      Assert.assertEquals(e0.getGroups(), e1.getGroups());
      Assert.assertEquals(e0.getID(), e1.getID());
      Assert.assertEquals(e0.getPublished(), e1.getPublished());
      Assert.assertEquals(e0.getPublisher(), e1.getPublisher());
      Assert.assertEquals(e0.getSummary(), e1.getSummary());
      Assert.assertEquals(e0.getThumbnail(), e1.getThumbnail());
      Assert.assertEquals(e0.getTitle(), e1.getTitle());
    }
  }
}

File: simplified-tests/src/main/java/org/nypl/simplified/tests/opds/OPDSJSONParserContract.kt
package org.nypl.simplified.tests.opds

import org.junit.Assert
import org.junit.Test
import org.nypl.simplified.books.core.BookFormats
import org.nypl.simplified.mime.MIMEType
import org.nypl.simplified.opds.core.OPDSAcquisition
import org.nypl.simplified.opds.core.OPDSAcquisitionFeed
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntry
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntryParser
import org.nypl.simplified.opds.core.OPDSAcquisitionFeedEntryParserType
import org.nypl.simplified.opds.core.OPDSAcquisitionPath
import org.nypl.simplified.opds.core.OPDSFeedParser
import org.nypl.simplified.opds.core.OPDSFeedParserType
import org.nypl.simplified.opds.core.OPDSJSONParser
import org.nypl.simplified.opds.core.OPDSJSONParserType
import org.nypl.simplified.opds.core.OPDSJSONSerializer
import org.nypl.simplified.opds.core.OPDSJSONSerializerType

import java.io.ByteArrayInputStream
import java.io.ByteArrayOutputStream
import java.io.FileNotFoundException
import java.io.InputStream
import java.net.URI
import java.net.URL

abstract class OPDSJSONParserContract {

  @Throws(Exception::class)
  private fun getResource(
    name: String): InputStream {

    val path = "/org/nypl/simplified/tests/opds/$name"
    val url = OPDSFeedEntryParserContract::class.java.getResource(path)
      ?: throw FileNotFoundException(path)
    return url.openStream()
  }

  @Test
  @Throws(Exception::class)
  fun testCompatibility20180921_1() {

    val json_parser = OPDSJSONParser.newParser()

    val e0 = json_parser.parseAcquisitionFeedEntryFromStream(
      getResource("compatibility-20180921-test-old.json"))

    val e1 = json_parser.parseAcquisitionFeedEntryFromStream(
      getResource("compatibility-20180921-test-new-1.json"))

    run {
      Assert.assertEquals(
        e0.acquisitionPaths.map(OPDSAcquisitionPath::finalContentType).map(MIMEType::fullType),
        e1.acquisitionPaths.map(OPDSAcquisitionPath::finalContentType).map(MIMEType::fullType))

      Assert.assertEquals(e0.acquisitions, e1.acquisitions)
      Assert.assertEquals(e0.availability, e1.availability)
      Assert.assertEquals(e0.authors, e1.authors)
      Assert.assertEquals(e0.categories, e1.categories)
      Assert.assertEquals(e0.cover, e1.cover)
      Assert.assertEquals(e0.groups, e1.groups)
      Assert.assertEquals(e0.id, e1.id)
      Assert.assertEquals(e0.published, e1.published)
      Assert.assertEquals(e0.publisher, e1.publisher)
      Assert.assertEquals(e0.summary, e1.summary)
      Assert.assertEquals(e0.thumbnail, e1.thumbnail)
      Assert.assertEquals(e0.title, e1.title)
    }
  }

  @Test
  @Throws(Exception::class)
  fun testCompatibility20180921_2() {

    val json_parser = OPDSJSONParser.newParser()

    val e0 = json_parser.parseAcquisitionFeedEntryFromStream(
      getResource("compatibility-20180921-test-old.json"))

    val e1 = json_parser.parseAcquisitionFeedEntryFromStream(
      getResource("compatibility-20180921-test-new-0.json"))

    run {
      Assert.assertEquals(
        e0.acquisitionPaths.map(OPDSAcquisitionPath::finalContentType).map(MIMEType::fullType),
        e1.acquisitionPaths.map(OPDSAcquisitionPath::finalContentType).map(MIMEType::fullType))

      Assert.assertEquals(e0.availability, e1.availability)
      Assert.assertEquals(e0.authors, e1.authors)
      Assert.assertEquals(e0.categories, e1.categories)
      Assert.assertEquals(e0.cover, e1.cover)
      Assert.assertEquals(e0.groups, e1.groups)
      Assert.assertEquals(e0.id, e1.id)
      Assert.assertEquals(e0.published, e1.published)
      Assert.assertEquals(e0.publisher, e1.publisher)
      Assert.assertEquals(e0.summary, e1.summary)
      Assert.assertEquals(e0.thumbnail, e1.thumbnail)
      Assert.assertEquals(e0.title, e1.title)
    }
  }
}
